// @ant-basement/base/dist/lib/constant/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Mode;
(function (Mode) {
    Mode["SERVER"] = "server";
    Mode["MINIAPP"] = "miniapp";
})(Mode = exports.Mode || (exports.Mode = {}));
var Method;
(function (Method) {
    Method["GET"] = "GET";
    Method["DEL"] = "DELETE";
    Method["PUT"] = "PUT";
    Method["POST"] = "POST";
})(Method = exports.Method || (exports.Method = {}));
var Service;
(function (Service) {
    Service["DB"] = "db";
    Service["FILE"] = "file";
    Service["FUNCTION"] = "function";
})(Service = exports.Service || (exports.Service = {}));
var FileEnv;
(function (FileEnv) {
    FileEnv["PUBLIC"] = "public";
    FileEnv["PRIVATE"] = "private";
})(FileEnv = exports.FileEnv || (exports.FileEnv = {}));
exports.HEADER_PREFIX = 'x-basement-';
exports.WHITELIST_EXTENSIONS = [
    'jpg',
    'jpeg',
    'png',
    'gif',
    'bmp',
    'wbmp',
    'webp',
    'bmp',
    'svg',
    'image',
    'au',
    'mp2',
    'mp3',
    '3gpp',
    'mp4',
    'mpeg',
    'flv',
    'avi',
    'wmv',
    'mpg',
    'mpga',
];
