// @ant-basement/base/dist/lib/basement.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const debug_1 = tslib_1.__importDefault(require("debug"));
const humanize_ms_1 = tslib_1.__importDefault(require("humanize-ms"));
const errors_1 = require("@ant-basement/errors");
const db_1 = tslib_1.__importDefault(require("./client/db"));
const file_1 = tslib_1.__importDefault(require("./client/file"));
const function_1 = tslib_1.__importDefault(require("./client/function"));
const constant_1 = require("./constant");
const utility_1 = require("./utility");
const logger = {
    info: debug_1.default('basement:info'),
    warn: debug_1.default('basement:warn'),
    error: debug_1.default('basement:error'),
};
class Basement {
    constructor(options) {
        utility_1.assert(options, 'options is required');
        utility_1.assert(options.mode, 'mode is required');
        utility_1.assert(options.spaceId, 'spaceId is required');
        utility_1.assert(options.endpoint, 'endpoint is required');
        utility_1.assert(options.httpclient, 'httpclient is required');
        this.logger = options.logger || logger;
        this.urllib = options.httpclient;
        delete options.logger;
        delete options.httpclient;
        this.options = options;
    }
    get db() {
        if (this._db) {
            return this._db;
        }
        this._db = new db_1.default(this);
        return this._db;
    }
    get file() {
        if (this._file) {
            return this._file;
        }
        this._file = new file_1.default(this);
        return this._file;
    }
    get ['function']() {
        if (this._function) {
            return this._function;
        }
        this._function = new function_1.default(this);
        return this._function;
    }
    createTraceId() {
        const digits = '1234567890abcdef';
        const length = digits.length;
        return (new Array(length)).fill(0).map(() => digits[Math.floor(Math.random() * length)]).join('');
    }
    async request(requestOptions) {
        this.logger.info('[REQUEST] requestOptions ' + JSON.stringify(requestOptions, null, 2));
        let params = {};
        params.query = Object.assign({ now: Date.now() }, requestOptions.query);
        const url = `${this.options.endpoint}${requestOptions.path}?${utility_1.queryToString(params.query)}`;
        if (requestOptions.stream) {
            params.stream = requestOptions.stream;
        }
        else if (requestOptions.data) {
            params.data = requestOptions.data;
        }
        params.method = requestOptions.method || constant_1.Method.POST;
        params.timeout = humanize_ms_1.default(this.options.timeout || requestOptions.timeout || '5s');
        params.headers = Object.assign({}, requestOptions.headers);
        params.dataType = 'json';
        params.contentType = 'json';
        params.serviceHeaders = Object.assign({}, requestOptions.serviceHeaders, {
            sdkType: this.options.mode,
            spaceId: this.options.spaceId,
            traceId: this.createTraceId(),
        });
        this.logger.info('[REQUEST] before transform ' + JSON.stringify(params, null, 2));
        params = this.transform(params);
        this.logger.info('[REQUEST] send request to url ' + url + ' with options ' + JSON.stringify(params, null, 2));
        const res = await this.urllib.request(url, params);
        if (res.status >= 400) {
            throw errors_1.composeFromHttpResponse(res.status, res.data);
        }
        return res.data;
    }
    toString() {
        return this.options;
    }
    transform(..._) {
        throw new errors_1.ECODE.UnsupportedError('should not use default implementation');
    }
}
exports.default = Basement;
