// @ant-basement/base/dist/lib/client/db.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const base_1 = tslib_1.__importDefault(require("./base"));
const utility_1 = require("../utility");
const constant_1 = require("../constant");
const utility_2 = require("@ant-basement/utility");
class DbClient extends base_1.default {
    constructor(builder) {
        super(builder, constant_1.Service.DB);
        this.path = '/gateway.execute';
    }
    async createCollection(name, options = {}) {
        const op = new utility_2.Op('createCollection', { name, options });
        return await this.request({
            path: this.path,
            data: op.inspect(),
        });
    }
    async collections() {
        const op = new utility_2.Op('collections');
        return await this.request({
            path: this.path,
            data: op.inspect(),
        });
    }
    collection(collection) {
        utility_1.assert(collection, 'collection name is required');
        const self = this;
        return new Proxy({}, {
            get(target, prop, receiver) {
                if (utility_2.isCollectionOp(prop)) {
                    return async function bound(...raw) {
                        prop = prop.toString();
                        const spec = utility_2.OP_SPECS[prop];
                        const args = spec.getArgsFromArray(raw || []);
                        args.collection = collection;
                        const op = new utility_2.Op(prop, args);
                        return utility_2.json2js(await self.request({
                            path: self.path,
                            data: op.inspect(),
                        }));
                    };
                }
                return Reflect.get(target, prop, receiver);
            },
        });
    }
}
exports.default = DbClient;
