// @ant-basement/base/dist/lib/client/file.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const base_1 = tslib_1.__importDefault(require("./base"));
const utility_1 = require("../utility");
const constant_1 = require("../constant");
const errors_1 = require("@ant-basement/errors");
class FileClient extends base_1.default {
    constructor(builder) {
        super(builder, constant_1.Service.FILE);
    }
    async deleteFile(url) {
        return await this.request({
            path: '/files/remove',
            data: { url },
        });
    }
    async getPrivateFileByUrl(url, expires) {
        return await this.request({
            path: '/files/sign',
            data: {
                url,
                expires,
            },
        });
    }
    async uploadFile(_) {
        throw new errors_1.ECODE.UnsupportedError('children implementation required');
    }
    validateFilename(name) {
        const extension = name.split('.').pop();
        utility_1.assert(constant_1.WHITELIST_EXTENSIONS.includes(extension), 'only image file are allowed');
    }
}
exports.default = FileClient;
