// @ant-basement/errors/lib/errors/control.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = require("../error");
class TenantNotAvailableError extends error_1.CommonError {
    constructor(message, errors) {
        super(null, message, errors);
    }
}
TenantNotAvailableError.code = 'TENANT_NOT_AVAILABLE';
TenantNotAvailableError.prefix = '租户未开通服务';
TenantNotAvailableError.httpStatusCode = 412;
exports.TenantNotAvailableError = TenantNotAvailableError;
class SpaceNotFoundError extends error_1.CommonError {
    constructor(message, errors) {
        super(null, message, errors);
    }
}
SpaceNotFoundError.code = 'SPACE_NOT_FOUND';
SpaceNotFoundError.prefix = 'Space 不存在';
SpaceNotFoundError.httpStatusCode = 404;
exports.SpaceNotFoundError = SpaceNotFoundError;
class ResourceNotInitializedError extends error_1.CommonError {
    constructor(message, errors) {
        super(null, message, errors);
    }
}
ResourceNotInitializedError.code = 'RESOURCE_NOT_INITIALIZED';
ResourceNotInitializedError.prefix = '资源未初始化';
ResourceNotInitializedError.httpStatusCode = 406;
exports.ResourceNotInitializedError = ResourceNotInitializedError;
class DecisionNotFoundError extends error_1.CommonError {
    constructor(message, errors) {
        super(null, message, errors);
    }
}
DecisionNotFoundError.code = 'DECISION_NOT_FOUND';
DecisionNotFoundError.prefix = '决策配置不存在';
DecisionNotFoundError.httpStatusCode = 404;
exports.DecisionNotFoundError = DecisionNotFoundError;
class SpaceNameConflictError extends error_1.CommonError {
    constructor(message, errors) {
        super(null, message, errors);
    }
}
SpaceNameConflictError.code = 'SPACE_NAME_CONFLICT';
SpaceNameConflictError.prefix = 'Space 名冲突';
SpaceNameConflictError.httpStatusCode = 409;
exports.SpaceNameConflictError = SpaceNameConflictError;
class ResourceMismatchError extends error_1.CommonError {
    constructor(message, errors) {
        super(null, message, errors);
    }
}
ResourceMismatchError.code = 'RESOURCE_MISMATCH';
ResourceMismatchError.prefix = '资源不匹配';
ResourceMismatchError.httpStatusCode = 409;
exports.ResourceMismatchError = ResourceMismatchError;
