// @ant-basement/utility/dist/lib/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const op_1 = tslib_1.__importDefault(require("./model/op"));
const utility_1 = require("./utility");
const specs_1 = require("./model/specs");
class Query {
    constructor(collection) {
        return new Proxy(this, {
            get(target, prop, receiver) {
                if (specs_1.isCollectionOp(prop)) {
                    return function bound(...raw) {
                        const op = prop.toString();
                        const spec = specs_1.OP_SPECS[op];
                        const args = spec.keys.reduce((accu, key, i) => {
                            accu[key] = raw[i];
                            return accu;
                        }, {});
                        args.collection = collection;
                        return new op_1.default(op, args);
                    };
                }
                return Reflect.get(target, prop, receiver);
            },
        });
    }
}
class Client {
    collection(name) {
        utility_1.validator.validate({ name: 'collectionName' }, { name });
        return new Query(name);
    }
    collections() {
        return new op_1.default('collections');
    }
    createCollection(name, options = {}) {
        return new op_1.default('createCollection', { name, options });
    }
}
exports.default = Client;
