// @ant-basement/utility/dist/lib/model/op.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("@ant-basement/errors");
const specs_1 = require("./specs");
const utility_1 = require("../utility");
const constant_1 = require("../constant");
class Op {
    constructor(op, args = {}, options = {}) {
        if (!specs_1.OP_SPECS.hasOwnProperty(op)) {
            const error = new errors_1.ECODE.NotFoundError('命令');
            error.setErrors(op);
            throw error;
        }
        this.op = op;
        this.spec = specs_1.OP_SPECS[op];
        for (const key of Object.keys(args)) {
            if (args[key] === undefined) {
                delete args[key];
            }
        }
        if (!args.options) {
            args.options = {};
        }
        this.args = utility_1.js2json(args);
        this.collection = args.collection;
        this.validate(this.args, options);
        if (options.format) {
            this.format();
        }
    }
    format() {
        delete this.args.options.readPreference;
        delete this.args.options.collation;
        delete this.args.options.session;
        delete this.args.options.cursor;
        delete this.args.options.w;
        delete this.args.options.j;
        delete this.args.options.wtimeout;
        delete this.args.options.serializeFunctions;
        if (this.spec.purpose === constant_1.PURPOSE.INDEX) {
            this.args.options.background = true;
        }
        switch (this.op) {
            case specs_1.OP_SPECS.COUNT.name:
            case specs_1.OP_SPECS.FIND.name:
            case specs_1.OP_SPECS.FIND_ONE.name: {
                this.args.options.limit = Math.min(constant_1.LIMIT, this.args.options.limit || Number.MAX_SAFE_INTEGER);
                break;
            }
            case specs_1.OP_SPECS.AGGREGATE.name: {
                this.args.options.cursor = {
                    batchSize: constant_1.LIMIT,
                };
                break;
            }
            default:
        }
        switch (this.op) {
            case specs_1.OP_SPECS.FIND.name:
            case specs_1.OP_SPECS.FIND_ONE.name:
            case specs_1.OP_SPECS.FIND_ONE_AND_UPDATE.name:
            case specs_1.OP_SPECS.FIND_ONE_AND_DELETE.name:
            case specs_1.OP_SPECS.FIND_ONE_AND_REPLACE.name:
            case specs_1.OP_SPECS.COUNT.name:
            case specs_1.OP_SPECS.DISTINCT.name:
            case specs_1.OP_SPECS.AGGREGATE.name:
                this.args.options.maxTimeMS = constant_1.QUERY_TIMEOUT;
                break;
            case specs_1.OP_SPECS.DELETE_ONE.name:
            case specs_1.OP_SPECS.DELETE_MANY.name:
            case specs_1.OP_SPECS.INSERT_ONE.name:
            case specs_1.OP_SPECS.INSERT_MANY.name:
            case specs_1.OP_SPECS.REPLACE_ONE.name:
            case specs_1.OP_SPECS.UPDATE_ONE.name:
            case specs_1.OP_SPECS.UPDATE_MANY.name:
                this.args.options.wtimeout = constant_1.QUERY_TIMEOUT;
                break;
            default:
        }
    }
    validate(args, options = {
        format: false,
        allowAdmin: false,
    }) {
        if (this.spec.scope === constant_1.SCOPE.COLLECTION && !this.collection) {
            throw new errors_1.ECODE.ValidationError('没有 collection 字段');
        }
        if (this.spec.purpose === constant_1.PURPOSE.ADMIN && !options.allowAdmin) {
            throw new errors_1.ECODE.ValidationError('不允许执行管理指令');
        }
        const errors = utility_1.validator.validate(this.spec.schema, args);
        if (errors) {
            const error = new errors_1.ECODE.ValidationError();
            error.setErrors(errors);
            throw error;
        }
    }
    inspect() {
        const args = Object.assign({}, this.args, {
            command: this.op,
        });
        if (Object.keys(args.options).length === 0) {
            delete args.options;
        }
        if (this.collection) {
            args.collection = this.collection;
        }
        return args;
    }
    getParams() {
        return this.spec.keys.map(key => this.args[key]);
    }
    hasPrefetch() {
        return [specs_1.OP_SPECS.INSERT_ONE.name, specs_1.OP_SPECS.INSERT_MANY.name].indexOf(this.op) > -1;
    }
    getPrefetch() {
        switch (this.op) {
            case specs_1.OP_SPECS.INSERT_ONE.name: {
                return {
                    affectedDocs: 1,
                    result: this.args.doc,
                };
            }
            case specs_1.OP_SPECS.INSERT_MANY.name: {
                return {
                    affectedDocs: this.args.docs.length,
                    result: this.args.docs,
                };
            }
            default:
        }
    }
}
exports.default = Op;
