// @ant-basement/utility/dist/lib/model/specs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const constant_1 = require("../constant");
const spec_1 = tslib_1.__importDefault(require("./spec"));
function snakeToCamel(key) {
    return key.toLowerCase().replace(/_(\w)/g, (_, match) => match.toUpperCase());
}
const specs = Object.assign({}, constant_1.DATABASE_OP, constant_1.COLLECTION_OP);
const OP_SPECS = {};
exports.OP_SPECS = OP_SPECS;
const OP_NAMES = [];
exports.OP_NAMES = OP_NAMES;
for (const key in specs) {
    if (specs.hasOwnProperty(key)) {
        const name = snakeToCamel(key);
        const spec = new spec_1.default(name, specs[key]);
        OP_SPECS[key] = spec;
        OP_SPECS[name] = spec;
        OP_NAMES.push(name);
    }
}
function isOpValid(op) {
    return typeof op === 'string' && OP_NAMES.indexOf(op) > -1;
}
exports.isOpValid = isOpValid;
function isCollectionOp(op) {
    return isOpValid(op) && OP_SPECS[op.toString()].scope === constant_1.SCOPE.COLLECTION;
}
exports.isCollectionOp = isCollectionOp;
function isDatabaseOp(op) {
    return isOpValid(op) && OP_SPECS[op.toString()].scope === constant_1.SCOPE.DATABASE;
}
exports.isDatabaseOp = isDatabaseOp;
