// @ant-basement/utility/dist/lib/constant/opSpec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const opMeta_1 = require("./opMeta");
const DATABASE_OP = {
    COLLECTIONS: {
        scope: opMeta_1.SCOPE.DATABASE,
        purpose: opMeta_1.PURPOSE.COLLECTION,
        operation: opMeta_1.OPERATION.READ,
    },
    CREATE_COLLECTION: {
        scope: opMeta_1.SCOPE.DATABASE,
        purpose: opMeta_1.PURPOSE.COLLECTION,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['name', 'collectionName'], ['options', 'object']]),
    },
    STATS: {
        scope: opMeta_1.SCOPE.DATABASE,
        purpose: opMeta_1.PURPOSE.ADMIN,
        operation: opMeta_1.OPERATION.READ,
    },
};
exports.DATABASE_OP = DATABASE_OP;
const COLLECTION_OP = {
    DROP: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.COLLECTION,
        operation: opMeta_1.OPERATION.WRITE,
    },
    RENAME: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.COLLECTION,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['newName', 'collectionName'], ['options', 'object']]),
    },
    CREATE_INDEX: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.INDEX,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['field', 'field'], ['options', 'object']]),
    },
    CREATE_INDEXES: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.INDEX,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['fields', 'array'], ['options', 'object']]),
    },
    LIST_INDEXES: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.INDEX,
        operation: opMeta_1.OPERATION.READ,
        signature: new Map([['options', 'object']]),
    },
    DROP_INDEX: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['indexName', 'field'], ['options', 'object']]),
    },
    DROP_INDEXES: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['options', 'object']]),
    },
    INSERT_ONE: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['doc', 'field'], ['options', 'object']]),
    },
    INSERT_MANY: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['docs', 'array'], ['options', 'object']]),
    },
    FIND: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.READ,
        signature: new Map([
            [
                'query',
                {
                    type: 'object',
                    required: false,
                },
            ],
            ['options', 'object'],
        ]),
    },
    FIND_ONE: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.READ,
        signature: new Map([['query', 'object'], ['options', 'object']]),
    },
    FIND_ONE_AND_UPDATE: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['filter', 'object'], ['update', 'object'], ['options', 'object']]),
    },
    FIND_ONE_AND_REPLACE: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['filter', 'object'], ['update', 'object'], ['options', 'object']]),
    },
    FIND_ONE_AND_DELETE: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['filter', 'object'], ['options', 'object']]),
    },
    UPDATE_ONE: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['filter', 'object'], ['update', 'object'], ['options', 'object']]),
    },
    UPDATE_MANY: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['filter', 'object'], ['update', 'object'], ['options', 'object']]),
    },
    REPLACE_ONE: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['doc', 'object'], ['filter', 'object'], ['options', 'object']]),
    },
    DELETE_ONE: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['filter', 'object'], ['options', 'object']]),
    },
    DELETE_MANY: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.WRITE,
        signature: new Map([['filter', 'object'], ['options', 'object']]),
    },
    COUNT: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.READ,
        signature: new Map([['query', 'object'], ['options', 'object']]),
    },
    DISTINCT: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.READ,
        signature: new Map([['key', 'string'], ['query', 'object'], ['options', 'object']]),
    },
    AGGREGATE: {
        scope: opMeta_1.SCOPE.COLLECTION,
        purpose: opMeta_1.PURPOSE.CRUD,
        operation: opMeta_1.OPERATION.READ,
        signature: new Map([['pipeline', 'object'], ['options', 'object']]),
    },
};
exports.COLLECTION_OP = COLLECTION_OP;
tslib_1.__exportStar(require("./opMeta"), exports);
