// @ant-basement/errors/lib/error.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
class CommonError extends Error {
    constructor(error, message, errors = {}) {
        super(message || (error && error.message) || '');
        this.errors = errors;
        this.reference = 'https://www.cloud.alipay.com/docs/2/91090';
        this.message = this.formatMessage(this.message);
        if (error) {
            if (error.stack) {
                this.stack = error.stack;
            }
            if (error.reference) {
                this.reference = error.reference;
            }
        }
        if (Error.hasOwnProperty('captureStackTrace')) {
            Error.captureStackTrace(this, CommonError);
        }
    }
    get httpStatusCode() {
        return this.constructor.httpStatusCode || 500;
    }
    formatMessage(message) {
        const parts = [];
        if (this.constructor.prefix) {
            parts.push(this.constructor.prefix);
        }
        if (message) {
            parts.push(message);
        }
        return parts.join('：');
    }
    setErrors(data) {
        this.errors = data;
    }
    inspect() {
        return {
            code: this.constructor.code,
            errors: this.errors,
            message: this.message,
            reference: this.reference,
        };
    }
}
CommonError.code = 'COMMON_ERROR';
exports.CommonError = CommonError;
