// @ant-basement/utility/dist/lib/utility/bson.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constant_1 = require("../constant");
exports.fromJs = {
    fromDate(val) {
        return val.toISOString();
    },
    fromRegexp(val) {
        return `/${val.source.replace(/\\\//g, '/')}/${val.flags}`;
    },
    fromObjectId(val) {
        return `${constant_1.PREFIXES.ObjectId}${val.toString()}`;
    },
};
exports.fromJson = {
    toDate(val) {
        return new Date(val);
    },
    toObjectId(val) {
        return val.slice(constant_1.PREFIXES.ObjectId.length);
    },
    toRegexp(val) {
        const firstIndex = val.indexOf('/');
        const lastIndex = val.lastIndexOf('/');
        return new RegExp(val.slice(firstIndex + 1, lastIndex), val.slice(lastIndex + 1));
    },
};
exports.is = {
    _isByProto(val, type) {
        return Object.prototype.toString.call(val) === `[object ${type}]`;
    },
    isArray(val) {
        return Array.isArray(val);
    },
    isObject(val) {
        return val !== null && typeof val === 'object' && !this.isArray(val);
    },
    isString(val) {
        return typeof val === 'string' || this._isByProto(val, 'String');
    },
    isDate(val) {
        return this.isString(val) && /^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/.test(val) && !isNaN(Date.parse(val));
    },
    isObjectId(val) {
        if (this.isString(val) && val.startsWith(constant_1.PREFIXES.ObjectId)) {
            const str = val.slice(constant_1.PREFIXES.ObjectId.length);
            if (str.length === constant_1.OID_LENGTH) {
                return true;
            }
        }
        return false;
    },
    isRegexp(val) {
        if (this.isString(val)) {
            let areFlagsValid = true;
            const parts = val.split('/');
            const last = parts[parts.length - 1];
            if (last) {
                areFlagsValid = last.split('').reduce((expr, p) => {
                    return expr === true && constant_1.REGEXP_FLAGS.indexOf(p) > -1;
                }, true);
            }
            return parts.length >= 2 && areFlagsValid;
        }
        return false;
    },
};
function js2json(data) {
    if (data instanceof RegExp) {
        return exports.fromJs.fromRegexp(data);
    }
    if (data instanceof Date) {
        return exports.fromJs.fromDate(data);
    }
    if (exports.is.isString(data) && /^[0-9a-fA-F]{24}$/.test(data)) {
        return exports.fromJs.fromObjectId(data);
    }
    if (exports.is.isArray(data)) {
        return data.map(js2json);
    }
    if (exports.is.isObject(data)) {
        return Object.keys(data).reduce((accu, k) => {
            accu[k] = js2json(data[k]);
            return accu;
        }, {});
    }
    return data;
}
exports.js2json = js2json;
function json2js(data) {
    if (exports.is.isRegexp(data)) {
        return exports.fromJson.toRegexp(data);
    }
    if (exports.is.isObjectId(data)) {
        return exports.fromJson.toObjectId(data);
    }
    if (exports.is.isDate(data)) {
        return exports.fromJson.toDate(data);
    }
    if (exports.is.isArray(data)) {
        return data.map(json2js);
    }
    if (exports.is.isObject(data)) {
        return Object.keys(data).reduce((accu, k) => {
            accu[k] = json2js(data[k]);
            return accu;
        }, {});
    }
    return data;
}
exports.json2js = json2js;
