// @ant-basement/utility/dist/lib/utility/validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("@ant-basement/errors");
const constant_1 = require("../constant");
function ruleOfField(_, value) {
    function validateField(name) {
        if (/[\.\$]/.test(name)) {
            return constant_1.MESSAGES.ILLEGAL;
        }
    }
    if (Array.isArray(value)) {
        return undefined;
    }
    if (typeof value === 'string') {
        return validateField(value);
    }
    let result = null;
    for (const key of Object.keys(value)) {
        result = validateField(key);
        if (typeof result === 'string') {
            return result;
        }
        if (typeof value[key] === 'object') {
            return ruleOfField(_, value[key]);
        }
    }
    if (result === null) {
        return constant_1.MESSAGES.INVALID_TYPE;
    }
}
exports.ruleOfField = ruleOfField;
function ruleOfFields(_, value) {
    if (!Array.isArray(value)) {
        return constant_1.MESSAGES.NOT_ARRAY;
    }
    let err;
    let i = 0;
    while (!err && i < value.length) {
        err = ruleOfField(null, value[i]);
        i++;
    }
    return err;
}
exports.ruleOfFields = ruleOfFields;
function ruleOfDatabaseName(_, value) {
    if (typeof value !== 'string') {
        return constant_1.MESSAGES.NOT_STRING;
    }
    if (value.trim().length === 0) {
        return constant_1.MESSAGES.NOT_EMPTY;
    }
    if (value.length > constant_1.MAX_DB_NAME_LENGTH) {
        return constant_1.MESSAGES.TOO_LONG;
    }
    if (/[\/\\\.\ \"\$]/.test(value)) {
        return constant_1.MESSAGES.ILLEGAL;
    }
    if (constant_1.INVALID_DB_NAMES.indexOf(value) > -1) {
        return constant_1.MESSAGES.SYSTEM;
    }
}
exports.ruleOfDatabaseName = ruleOfDatabaseName;
function ruleOfCollectionName(_, value) {
    if (typeof value !== 'string') {
        return constant_1.MESSAGES.NOT_STRING;
    }
    if (value.trim().length === 0) {
        return constant_1.MESSAGES.NOT_EMPTY;
    }
    if (value.length > constant_1.MAX_ID_NAME_LENGTH) {
        return constant_1.MESSAGES.TOO_LONG;
    }
    if (value.indexOf('$') > -1) {
        return constant_1.MESSAGES.ILLEGAL;
    }
    if (value.startsWith('system.')) {
        return constant_1.MESSAGES.ILLEGAL;
    }
    if (!/^[_a-z]/i.test(value)) {
        return constant_1.MESSAGES.ILLEGAL;
    }
}
exports.ruleOfCollectionName = ruleOfCollectionName;
const validator = new errors_1.Validator();
exports.validator = validator;
validator.addRule('field', ruleOfField);
validator.addRule('fields', ruleOfFields);
validator.addRule('databaseName', ruleOfDatabaseName);
validator.addRule('collectionName', ruleOfCollectionName);
